src/objdump.py

```diff
@@ -150,10 +162,104 @@ def parse_relocated_line(line: str) -> Tuple[str, str, str]:
     return before, imm, after
 
 
+def pre_process(mnemonic: str, args: str, next_row: Optional[str]) -> Tuple[str, str]:
+
+    if next_row and "R_PPC_EMB_SDA21" in next_row:
+        # With sda21 relocs, the linker transforms `r0` into `r2`/`r13`, and
+        # we may encounter this in either pre-transformed or post-transformed
+        # versions depending on if the .o file comes from compiler output or
+        # from disassembly. Normalize, to make sure both forms are treated as
+        # equivalent.
+        args = args.replace("(r2)", "(0)")
+        args = args.replace("(r13)", "(0)")
+        args = args.replace(",r2,", ",0,")
+        args = args.replace(",r13,", ",0,")
+
+        # We want to convert li and lis with an sda21 reloc,
+        # because the r0 to r2/r13 transformation results in
+        # turning an li/lis into an addi/addis with r2/r13 arg
+        # our preprocessing normalizes all versions to addi with a 0 arg
+
+        if mnemonic in {"li", "lis"}:
+            mnemonic = mnemonic.replace("li", "addi")
+            args_parts = args.split(",")
+            args = args_parts[0] + ",0," + args_parts[1]
+
+    return mnemonic, args
+
+
+def process_reloc(reloc_row: str, prev: str) -> Optional[str]:
+    if prev == "<skipped>":
+        return None
+
+    before, imm, after = parse_relocated_line(prev)
+    repl = reloc_row.split()[-1]
+    # As part of ignoring branch targets, we ignore relocations for j
+    # instructions. The target is already lost anyway.
+    if imm == "<target>":
+        assert ign_branch_targets
+        return None
+
+    if "R_PPC_" in reloc_row:
+
+        assert any(
+            r in reloc_row for r in ["R_PPC_REL24", "R_PPC_ADDR16", "R_PPC_EMB_SDA21"]
+        ), f"unknown relocation type '{reloc_row}' for line '{prev}'"
+
+        if "R_PPC_REL24" in reloc_row:
+            # function calls
+            pass
+        elif "R_PPC_ADDR16_HI" in reloc_row:
+            # absolute hi of addr
+            repl = f"{repl}@h"  ## not actually needed by permutor
```

Comment:
Permut**e**r, but also, aren't these markers actually used by the scorer?

No need for duplicated comments in all the cases, at least. If we do need a comment it could go at the top of this function and speak in general terms, because I think the same applies to MIPS as for PPC

Reply:
atleast for PPC it's not needed because we do the function `field_matches_any_symbol` and doesn't care if it has a @ ha suffix

or otherwise it just checks if lhs==rhs.. so we can either add the suffix to both sides or neither side.. both ways they match.

also I have the debug mode now, which prints the asm, so I think its helpful to see the @sda21|ha there.  and also just makes the code more similar to asm-differ 


I am moving the comment at the top of the function, I am not certain about MIPS if its the same or not.  Or maybe I should just remove the comment all together

Reply:
yea I am deciding to just remove the comment all together, but let me know if you think one should be there